src/lib/telemetry/TelemetryReporter.ts

```diff
@@ -1,13 +1,47 @@
+import { readdir, readJson, remove, writeJson, ensureDir, ensureFile, move } from 'fs-extra'
+import { randomBytes } from 'crypto'
+import { isArray } from 'util'
+import * as lockfile from 'lockfile'
+import { join, basename } from 'path'
+
 import { TelemetryClient } from '../../clients/telemetryClient'
 import { region } from '../../env'
 import userAgent from '../../user-agent'
 import { createIOContext, createTelemetryClient } from '../clients'
 import { SessionManager } from '../session/SessionManager'
 import { TelemetryLocalStore } from './TelemetryStore'
+import { ErrorReport } from '../error/ErrorReport'
+import { TelemetryCollector } from './TelemetryCollector'
+import { ErrorCodes } from '../error/ErrorCodes'
+
+class FileLock {
+  public readonly lockName: string
+  constructor(private lockPath: string, private options: any) {
+    this.lockName = basename(lockPath)
+  }
+
+  public async lock() {
+    await ensureDir(TelemetryReporter.PENDING_DATA_DIR)
+    return new Promise((resolve, reject) => {
+      lockfile.lock(this.lockPath, this.options, err => {
+        err ? reject(err) : resolve()
+      })
+    })
+  }
+
+  public unlock() {
+    try {
+      lockfile.unlock(this.lockPath)
```

Comment:
I think `lockfile.unlock` works the same way that `lock` works, it receives an callback, so we could wrap it inside a promise:
https://www.npmjs.com/package/lockfile#lockfileunlockpath-cb
```
lockFile.lock(path, [opts], cb)
Acquire a file lock on the specified path

lockFile.lockSync(path, [opts])
Acquire a file lock on the specified path

lockFile.unlock(path, cb)
Close and unlink the lockfile.
```